server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -776,6 +789,35 @@ public ClusterState execute(BatchExecutionContext<StartedShardUpdateTask> batchE
             return maybeUpdatedState;
         }
 
+        private ClusterState maybeRemoveIndexRefreshBlocks(
+            ClusterState clusterState,
+            @Nullable Set<Index> indicesWithUnpromotableShardsStarted
+        ) {
+            if (indicesWithUnpromotableShardsStarted == null) {
+                return clusterState;
+            }
+
+            ClusterBlocks.Builder clusterBlocksBuilder = null;
+            for (Index indexWithUnpromotableShardsStarted : indicesWithUnpromotableShardsStarted) {
+                String indexName = indexWithUnpromotableShardsStarted.getName();
+                assert clusterState.blocks().hasIndexBlock(indexName, INDEX_REFRESH_BLOCK);
+
+                var indexRoutingTable = clusterState.routingTable().index(indexWithUnpromotableShardsStarted);
+                if (indexRoutingTable.readyForSearch(clusterState)) {
```

Comment:
TIL about `readyForSearch` but apparently it does not use the `clusterState` parameter... looks like a wrong refactoring or a bug.

Reply:
Yes, I noticed that too, I'm planning on removing the parameter in a follow-up PR, the current logic seems correct to me.